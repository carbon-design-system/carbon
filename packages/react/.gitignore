# Storybook
storybook-static

# Generated metadata
react-docgen.json

# Cypress artifacts
cypress/videos
cypress/screenshots