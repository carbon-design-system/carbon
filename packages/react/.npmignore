**/__mocks__/**
**/__tests__/**
**/examples/**
**/tasks/**
**/*-test.js
