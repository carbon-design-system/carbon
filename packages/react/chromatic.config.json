{
  "onlyChanged": true,
  "projectId": "Project:5e5429148f3c9a0022a380c4",
  "buildScriptName": "storybook:build",
  "storybookBaseDir": "packages/react",
  "zip": true
}
