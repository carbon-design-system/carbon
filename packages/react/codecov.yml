coverage:
  status:
    project:
      components:
        target: 100%
        threshold: 1%
        paths:
          - src/components/
