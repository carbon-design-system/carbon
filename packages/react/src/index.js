/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export Accordion from './components/Accordion';
export AccordionItem from './components/AccordionItem';
export { AspectRatio } from './components/AspectRatio';
export { Breadcrumb, BreadcrumbItem } from './components/Breadcrumb';
export Button from './components/Button';
export ButtonSet from './components/ButtonSet';
export Checkbox from './components/Checkbox';
export CodeSnippet from './components/CodeSnippet';
export ComboBox from './components/ComboBox';
export ComposedModal, {
  ModalHeader,
  ModalBody,
  ModalFooter,
} from './components/ComposedModal';
export ContentSwitcher from './components/ContentSwitcher';
export Copy from './components/Copy';
export CopyButton from './components/CopyButton';
export DangerButton from './components/DangerButton';
export {
  default as DataTable,
  Table,
  TableActionList,
  TableBatchAction,
  TableBatchActions,
  TableBody,
  TableCell,
  TableContainer,
  TableExpandHeader,
  TableExpandRow,
  TableExpandedRow,
  TableHead,
  TableHeader,
  TableRow,
  TableSelectAll,
  TableSelectRow,
  TableToolbar,
  TableToolbarAction,
  TableToolbarContent,
  TableToolbarSearch,
  TableToolbarMenu,
} from './components/DataTable';
export DatePicker from './components/DatePicker';
export DatePickerInput from './components/DatePickerInput';
export Dropdown from './components/Dropdown';
export {
  ErrorBoundary,
  ErrorBoundaryContext,
} from './components/ErrorBoundary';
export FileUploader, {
  Filename,
  FileUploaderButton,
  FileUploaderDropContainer,
  FileUploaderItem,
} from './components/FileUploader';
export Form from './components/Form';
export FluidForm from './components/FluidForm';
export FormGroup from './components/FormGroup';
export FormItem from './components/FormItem';
export FormLabel from './components/FormLabel';
export { Grid, Row, Column } from './components/Grid';
export Icon from './components/Icon';
export InlineLoading from './components/InlineLoading';
export Link from './components/Link';
export ListItem from './components/ListItem';
export Loading from './components/Loading';
export Modal from './components/Modal';
export ModalWrapper from './components/ModalWrapper';
export MultiSelect from './components/MultiSelect';
export {
  ToastNotification,
  InlineNotification,
  NotificationActionButton,
  NotificationButton,
  NotificationTextDetails,
} from './components/Notification';
export NumberInput from './components/NumberInput';
export OrderedList from './components/OrderedList';
export OverflowMenu from './components/OverflowMenu';
export OverflowMenuItem from './components/OverflowMenuItem';
export Pagination from './components/Pagination';
export PaginationNav from './components/PaginationNav';
export PrimaryButton from './components/PrimaryButton';
export {
  ProgressIndicator,
  ProgressStep,
} from './components/ProgressIndicator';
export RadioButton from './components/RadioButton';
export RadioButtonGroup from './components/RadioButtonGroup';
export Search from './components/Search';
export SearchFilterButton from './components/SearchFilterButton';
export SearchLayoutButton from './components/SearchLayoutButton';
export ExpandableSearch from './components/ExpandableSearch';
export SecondaryButton from './components/SecondaryButton';
export Select from './components/Select';
export SelectItem from './components/SelectItem';
export SelectItemGroup from './components/SelectItemGroup';
export Switch from './components/Switch';
export Slider from './components/Slider';
export {
  StructuredListWrapper,
  StructuredListHead,
  StructuredListBody,
  StructuredListRow,
  StructuredListInput,
  StructuredListCell,
  StructuredListSkeleton,
} from './components/StructuredList';
export Tab from './components/Tab';
export TabContent from './components/TabContent';
export Tabs from './components/Tabs';
export Tag from './components/Tag';
export TextArea from './components/TextArea';
export TextInput from './components/TextInput';
export {
  Tile,
  ClickableTile,
  SelectableTile,
  ExpandableTile,
  TileAboveTheFoldContent,
  TileBelowTheFoldContent,
} from './components/Tile';
export RadioTile from './components/RadioTile';
export TileGroup from './components/TileGroup';
export TimePicker from './components/TimePicker';
export TimePickerSelect from './components/TimePickerSelect';
export Toggle from './components/Toggle';
export ToggleSmall from './components/ToggleSmall';
export Toolbar, {
  ToolbarItem,
  ToolbarTitle,
  ToolbarOption,
  ToolbarDivider,
} from './components/Toolbar';
export ToolbarSearch from './components/ToolbarSearch';
export Tooltip from './components/Tooltip';
export TooltipDefinition from './components/TooltipDefinition';
export TooltipIcon from './components/TooltipIcon';
export UnorderedList from './components/UnorderedList';
export SkeletonText from './components/SkeletonText';
export SkeletonPlaceholder from './components/SkeletonPlaceholder';
export DataTableSkeleton from './components/DataTableSkeleton';
export AccordionSkeleton from './components/Accordion/Accordion.Skeleton';
export BreadcrumbSkeleton from './components/Breadcrumb/Breadcrumb.Skeleton';
export ButtonSkeleton from './components/Button/Button.Skeleton';
export CheckboxSkeleton from './components/Checkbox/Checkbox.Skeleton';
export CodeSnippetSkeleton from './components/CodeSnippet/CodeSnippet.Skeleton';
export DropdownSkeleton from './components/Dropdown/Dropdown.Skeleton';
export FileUploaderSkeleton from './components/FileUploader/FileUploader.Skeleton';
export NumberInputSkeleton from './components/NumberInput/NumberInput.Skeleton';
export PaginationSkeleton from './components/Pagination/Pagination.Skeleton';
export ProgressIndicatorSkeleton from './components/ProgressIndicator/ProgressIndicator.Skeleton';
export RadioButtonSkeleton from './components/RadioButton/RadioButton.Skeleton';
export SearchSkeleton from './components/Search/Search.Skeleton';
export SelectSkeleton from './components/Select/Select.Skeleton';
export SliderSkeleton from './components/Slider/Slider.Skeleton';
export TabsSkeleton from './components/Tabs/Tabs.Skeleton';
export TagSkeleton from './components/Tag/Tag.Skeleton';
export TextAreaSkeleton from './components/TextArea/TextArea.Skeleton';
export TextInputSkeleton from './components/TextInput/TextInput.Skeleton';
export ToggleSkeleton from './components/Toggle/Toggle.Skeleton';
export ToggleSmallSkeleton from './components/ToggleSmall/ToggleSmall.Skeleton';
export IconSkeleton from './components/Icon/Icon.Skeleton';
export DatePickerSkeleton from './components/DatePicker/DatePicker.Skeleton';
export {
  Content,
  Header,
  HeaderContainer,
  HeaderGlobalAction,
  HeaderGlobalBar,
  HeaderMenu,
  HeaderMenuButton,
  HeaderMenuItem,
  HeaderName,
  HeaderNavigation,
  HeaderPanel,
  HeaderSideNavItems,
  Switcher,
  SwitcherItem,
  SwitcherDivider,
  SkipToContent,
  SideNav,
  SideNavDetails,
  SideNavDivider,
  SideNavFooter,
  SideNavHeader,
  SideNavIcon,
  SideNavItem,
  SideNavItems,
  SideNavLink,
  SideNavLinkText,
  SideNavMenu,
  SideNavMenuItem,
  SideNavSwitcher,
} from './components/UIShell';

// Experimental
export {
  FeatureFlags as unstable_FeatureFlags,
  useFeatureFlag as unstable_useFeatureFlag,
  useFeatureFlags as unstable_useFeatureFlags,
} from './components/FeatureFlags';
export {
  PageSelector as unstable_PageSelector,
  Pagination as unstable_Pagination,
} from './components/Pagination/experimental';
export unstable_TreeView, {
  TreeNode as unstable_TreeNode,
} from './components/TreeView';
export unstable_Menu, {
  MenuDivider as unstable_MenuDivider,
  MenuGroup as unstable_MenuGroup,
  MenuItem as unstable_MenuItem,
  MenuRadioGroup as unstable_MenuRadioGroup,
  MenuSelectableItem as unstable_MenuSelectableItem,
} from './components/Menu';
export { useContextMenu as unstable_useContextMenu } from './components/ContextMenu';
export {
  Heading as unstable_Heading,
  Section as unstable_Section,
} from './components/Heading';
export { default as unstable_ProgressBar } from './components/ProgressBar';
