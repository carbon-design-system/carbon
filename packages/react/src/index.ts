/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import './feature-flags';

export * from './components/Accordion';
export * from './components/AccordionItem';
export * from './components/AspectRatio';
export * from './components/Breadcrumb';
export * from './components/Button';
export * from './components/ButtonSet';
export * from './components/Checkbox';
export * from './components/ClassPrefix';
export * from './components/CodeSnippet';
export * from './components/ComboBox';
export * from './components/ComposedModal';
export * from './components/ContainedList';
export * from './components/ContentSwitcher';
export * from './components/Copy';
export * from './components/CopyButton';
export * from './components/DangerButton';
export * from './components/DataTable';
export * from './components/DatePicker';
export * from './components/DatePickerInput';
export * from './components/Dropdown';
export * from './components/ErrorBoundary';
export * from './components/FileUploader';
export * from './components/FilterableMultiSelect';
export * from './components/Form';
export * from './components/FluidForm';
export * from './components/FormGroup';
export * from './components/FormItem';
export * from './components/FormLabel';
export * from './components/Grid';
export * from './components/IdPrefix';
export * from './components/InlineLoading';
export * from './components/Link';
export * from './components/ListItem';
export * from './components/Loading';
export * from './components/Modal';
export * from './components/ModalWrapper';
export * from './components/MultiSelect';
export * from './components/Notification';
export * from './components/NumberInput';
export * from './components/OrderedList';
export * from './components/OverflowMenu';
export * from './components/OverflowMenuItem';
export * from './components/Pagination';
export * from './components/PaginationNav';
export * from './components/PasswordInput';
export * from './components/PrimaryButton';
export * from './components/ProgressIndicator';
export * from './components/RadioButton';
export * from './components/RadioButtonGroup';
export * from './components/Search';
export * from './components/ExpandableSearch';
export * from './components/SecondaryButton';
export * from './components/Select';
export * from './components/SelectItem';
export * from './components/SelectItemGroup';
export * from './components/Switch';
export * from './components/Slider';
export * from './components/StructuredList';
export * from './components/Tab';
export * from './components/TabContent';
export * from './components/Tabs';
export * from './components/Tag';
export * from './components/TextArea';
export * from './components/TextInput';
export * from './components/Tile';
export * from './components/RadioTile';
export * from './components/TileGroup';
export * from './components/TimePicker';
export * from './components/TimePickerSelect';
export * from './components/Toggle';
export * from './components/Toggletip';
export * from './components/TreeView';
export * from './components/UnorderedList';
export * from './components/SkeletonText';
export * from './components/SkeletonPlaceholder';
export * from './components/SkeletonIcon';
export * from './components/DataTableSkeleton';
export * from './components/Pagination/Pagination.Skeleton';
export * from './components/RadioButton/RadioButton.Skeleton';
export * from './components/Tag/Tag.Skeleton';
export * from './components/Toggle/Toggle.Skeleton';
export * from './components/ToggleSmall/ToggleSmall.Skeleton';
export * from './components/Icon/Icon.Skeleton';
export * from './components/UIShell';

// Experimental
export { useContextMenu as unstable_useContextMenu } from './components/ContextMenu';
export {
  FeatureFlags as unstable_FeatureFlags,
  useFeatureFlag as unstable_useFeatureFlag,
  useFeatureFlags as unstable_useFeatureFlags,
} from './components/FeatureFlags';
export {
  FluidComboBox as unstable__FluidComboBox,
  FluidComboBoxSkeleton as unstable__FluidComboBoxSkeleton,
} from './components/FluidComboBox';
export {
  FluidDatePicker as unstable__FluidDatePicker,
  FluidDatePickerSkeleton as unstable__FluidDatePickerSkeleton,
} from './components/FluidDatePicker';
export { FluidDatePickerInput as unstable__FluidDatePickerInput } from './components/FluidDatePickerInput';
export {
  FluidDropdown as unstable__FluidDropdown,
  FluidDropdownSkeleton as unstable__FluidDropdownSkeleton,
} from './components/FluidDropdown';
export {
  FluidMultiSelect as unstable__FluidMultiSelect,
  FluidMultiSelectSkeleton as unstable__FluidMultiSelectSkeleton,
} from './components/FluidMultiSelect';
export {
  FluidSelect as unstable__FluidSelect,
  FluidSelectSkeleton as unstable__FluidSelectSkeleton,
} from './components/FluidSelect';
export {
  FluidTextArea as unstable__FluidTextArea,
  FluidTextAreaSkeleton as unstable__FluidTextAreaSkeleton,
} from './components/FluidTextArea';
export {
  FluidTextInput as unstable__FluidTextInput,
  FluidTextInputSkeleton as unstable__FluidTextInputSkeleton,
} from './components/FluidTextInput';
export {
  FluidTimePicker as unstable__FluidTimePicker,
  FluidTimePickerSkeleton as unstable__FluidTimePickerSkeleton,
} from './components/FluidTimePicker';
export { FluidTimePickerSelect as unstable__FluidTimePickerSelect } from './components/FluidTimePickerSelect';
export * from './components/Heading';
export * from './components/IconButton';
export * from './components/Layer';
export {
  LayoutDirection as unstable_LayoutDirection,
  useLayoutDirection as unstable_useLayoutDirection,
} from './components/Layout';
export {
  Menu as unstable_Menu,
  MenuItem as unstable_MenuItem,
  MenuItemDivider as unstable_MenuItemDivider,
  MenuItemGroup as unstable_MenuItemGroup,
  MenuItemRadioGroup as unstable_MenuItemRadioGroup,
  MenuItemSelectable as unstable_MenuItemSelectable,
} from './components/Menu';
export { OverflowMenuV2 as unstable_OverflowMenuV2 } from './components/OverflowMenuV2';
export {
  PageSelector as unstable_PageSelector,
  Pagination as unstable_Pagination,
} from './components/Pagination/experimental';
export * from './components/Popover';
export * from './components/ProgressBar';
export * from './components/Stack';
export * from './components/Tooltip';
export {
  Text as unstable_Text,
  TextDirection as unstable_TextDirection,
} from './components/Text';
export * from './components/Tooltip/DefinitionTooltip';
export * from './components/Theme';
export * from './internal/usePrefix';
export { useIdPrefix } from './internal/useIdPrefix';
