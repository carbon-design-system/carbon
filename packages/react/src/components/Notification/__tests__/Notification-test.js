/**
 * Copyright IBM Corp. 2016, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { ErrorFilled } from '@carbon/icons-react';
import {
  NotificationButton,
  ToastNotification,
  InlineNotification,
  ActionableNotification,
  StaticNotification,
  Callout,
} from '../Notification';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const prefix = 'cds';

describe('NotificationButton', () => {
  it('should place the `className` prop on the outermost DOM node', () => {
    const { container } = render(<NotificationButton className="test" />);
    expect(container.firstChild).toHaveClass('test');
  });

  it('supports custom icon', () => {
    const { rerender } = render(<NotificationButton />);
    const defaultIcon = screen.queryByRole('button').innerHTML;

    rerender(
      <NotificationButton
        renderIcon={(props) => <ErrorFilled size={20} {...props} />}
      />
    );
    const customIcon = screen.queryByRole('button').innerHTML;

    expect(defaultIcon).not.toEqual(customIcon);
  });

  it('interpolates matching className based on notificationType prop', () => {
    const { rerender, container } = render(<NotificationButton />);
    const notificationTypes = ['toast', 'inline'];

    notificationTypes.forEach((notificationType) => {
      rerender(<NotificationButton notificationType={notificationType} />);
      expect(container.firstChild).toHaveClass(
        `${prefix}--${notificationType}-notification__close-button`
      );
      expect(screen.queryByRole('button').firstChild).toHaveClass(
        `${prefix}--${notificationType}-notification__close-icon`
      );
    });
  });
});

describe('ToastNotification', () => {
  it('should render', () => {
    const { asFragment } = render(
      <ToastNotification title="Notification title" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should have role=status by default', () => {
    const { container } = render(
      <ToastNotification title="Notification title" />
    );
    expect(container.firstChild).toHaveAttribute('role', 'status');
  });

  it('should place the `className` prop on the outermost DOM node', () => {
    const { container } = render(
      <ToastNotification title="Notification title" className="test" />
    );
    expect(container.firstChild).toHaveClass('test');
  });

  it('interpolates matching className based on kind prop', () => {
    const { rerender } = render(
      <ToastNotification title="Notification title" />
    );
    const kinds = [
      'error',
      'info',
      'info-square',
      'success',
      'warning',
      'warning-alt',
    ];
    kinds.forEach((kind) => {
      rerender(<ToastNotification kind={kind} />);
      expect(screen.queryByRole('status')).toHaveClass(
        `${prefix}--toast-notification--${kind}`
      );
    });
  });

  it('allows non-interactive elements as children', () => {
    render(
      <ToastNotification title="Notification title">
        <p>Sample text</p>
      </ToastNotification>
    );
    // eslint-disable-next-line testing-library/prefer-presence-queries
    expect(screen.queryByText(/Sample text/i)).toBeInTheDocument();
  });

  it('does not allow interactive elements as children', () => {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {});

    expect(() => {
      render(
        <ToastNotification title="Notification title">
          <button type="button">Sample button text</button>
        </ToastNotification>
      );
    }).toThrow();

    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  it('does not render close button when `hideCloseButton` is provided', () => {
    render(<ToastNotification title="Notification title" hideCloseButton />);
    const closeButton = screen.queryByRole('button', {
      hidden: true,
    });
    expect(closeButton).not.toBeInTheDocument();
  });

  it('calls `onClose` when notification is closed', async () => {
    const onClose = jest.fn();
    render(<ToastNotification title="Notification title" onClose={onClose} />);

    const closeButton = screen.queryByRole('button', {
      hidden: true,
    });
    await userEvent.click(closeButton);
    expect(onClose).toHaveBeenCalledTimes(1);
    await waitFor(() => {
      expect(screen.queryByRole('status')).not.toBeInTheDocument();
    });
  });

  it('keeps notification open if `onClose` returns false', async () => {
    render(
      <ToastNotification title="Notification title" onClose={() => false} />
    );

    const closeButton = screen.queryByRole('button', {
      hidden: true,
    });
    await userEvent.click(closeButton);
    // eslint-disable-next-line testing-library/prefer-presence-queries
    expect(screen.queryByRole('status')).toBeInTheDocument();
  });

  it('calls `onCloseButtonClick` when notification is closed', async () => {
    const onCloseButtonClick = jest.fn();
    render(
      <ToastNotification
        title="Notification title"
        onCloseButtonClick={onCloseButtonClick}
      />
    );

    const closeButton = screen.queryByRole('button', {
      hidden: true,
    });
    await userEvent.click(closeButton);
    expect(onCloseButtonClick).toHaveBeenCalledTimes(1);
  });
});

describe('InlineNotification', () => {
  it('should render', () => {
    const { asFragment } = render(
      <InlineNotification title="Notification title" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should have role=status by default', () => {
    const { container } = render(
      <InlineNotification title="Notification title" />
    );
    expect(container.firstChild).toHaveAttribute('role', 'status');
  });

  it('should place the `className` prop on the outermost DOM node', () => {
    const { container } = render(
      <InlineNotification title="Notification title" className="test" />
    );
    expect(container.firstChild).toHaveClass('test');
  });

  it('interpolates matching className based on kind prop', () => {
    const { rerender } = render(
      <InlineNotification title="Notification title" />
    );
    const kinds = [
      'error',
      'info',
      'info-square',
      'success',
      'warning',
      'warning-alt',
    ];
    kinds.forEach((kind) => {
      rerender(<InlineNotification title="Notification title" kind={kind} />);
      expect(screen.queryByRole('status')).toHaveClass(
        `${prefix}--inline-notification--${kind}`
      );
    });
  });

  it('allows non-interactive elements as children', () => {
    render(
      <InlineNotification title="Notification title">
        <p>Sample text</p>
      </InlineNotification>
    );
    // eslint-disable-next-line testing-library/prefer-presence-queries
    expect(screen.queryByText(/Sample text/i)).toBeInTheDocument();
  });

  it('does not allow interactive elements as children', () => {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {});

    expect(() => {
      render(
        <InlineNotification title="Notification title">
          <button type="button">Sample button text</button>
        </InlineNotification>
      );
    }).toThrow();

    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  it('does not render close button when `hideCloseButton` is provided', () => {
    render(<InlineNotification title="Notification title" hideCloseButton />);
    const closeButton = screen.queryByRole('button', {
      hidden: true,
    });
    expect(closeButton).not.toBeInTheDocument();
  });

  it('calls `onClose` when notification is closed', async () => {
    const onClose = jest.fn();
    render(<InlineNotification title="Notification title" onClose={onClose} />);

    const closeButton = screen.queryByRole('button', {
      hidden: true,
    });
    await userEvent.click(closeButton);
    expect(onClose).toHaveBeenCalledTimes(1);
    await waitFor(() => {
      expect(screen.queryByRole('status')).not.toBeInTheDocument();
    });
  });

  it('keeps notification open if `onClose` returns false', async () => {
    render(
      <InlineNotification title="Notification title" onClose={() => false} />
    );

    const closeButton = screen.queryByRole('button', {
      hidden: true,
    });
    await userEvent.click(closeButton);
    // eslint-disable-next-line testing-library/prefer-presence-queries
    expect(screen.queryByRole('status')).toBeInTheDocument();
  });

  it('calls `onCloseButtonClick` when notification is closed', async () => {
    const onCloseButtonClick = jest.fn();
    render(
      <InlineNotification
        title="Notification title"
        onCloseButtonClick={onCloseButtonClick}
      />
    );

    const closeButton = screen.queryByRole('button', {
      hidden: true,
    });
    await userEvent.click(closeButton);
    expect(onCloseButtonClick).toHaveBeenCalledTimes(1);
  });
});

describe('ActionableNotification', () => {
  it('should render', () => {
    const { asFragment } = render(
      <ActionableNotification actionButtonLabel="My custom action" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('uses role=alertdialog', () => {
    const { container } = render(
      <ActionableNotification actionButtonLabel="My custom action" />
    );

    expect(container.firstChild).toHaveAttribute('role', 'alertdialog');
  });

  it('renders correct action label', () => {
    render(<ActionableNotification actionButtonLabel="My custom action" />);
    const actionButton = screen.queryByRole('button', {
      name: 'My custom action',
    });
    expect(actionButton).toBeInTheDocument();
  });

  it('closes notification via escape button when focus is placed on the notification', async () => {
    const onCloseButtonClick = jest.fn();
    const onClose = jest.fn();
    render(
      <ActionableNotification
        onClose={onClose}
        onCloseButtonClick={onCloseButtonClick}
        actionButtonLabel="My custom action"
      />
    );

    // after focus is placed, the notification should close via escape
    await userEvent.keyboard('{Escape}');
    expect(onCloseButtonClick).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledTimes(1);

    await waitFor(() => {
      expect(screen.queryByRole('alertdialog')).not.toBeInTheDocument();
    });
  });
});

// TODO: Remove StaticNotification tests when StaticNotification is removed (v12)
describe('StaticNotification', () => {
  it('logs a deprecation notice when used', () => {
    const spy = jest.spyOn(console, 'warn').mockImplementation(() => {});

    expect(() => {
      render(<StaticNotification title="Notification title" />);
    }).not.toThrow();

    expect(spy).toHaveBeenCalledWith(
      'Warning: `StaticNotification` has been renamed to `Callout`.' +
        'Run the following codemod to automatically update usages in your' +
        'project: `npx @carbon/upgrade migrate refactor-to-callout --write`'
    );
    spy.mockRestore();
  });
});

describe('Callout', () => {
  it('should place the `className` prop on the outermost DOM node', () => {
    const { container } = render(<Callout className="test" />);
    expect(container.firstChild).toHaveClass('test');
  });

  it('calls `onActionButtonClick` when action button is clicked', async () => {
    const onActionButtonClick = jest.fn();
    render(
      <Callout
        title="Notification title"
        actionButtonLabel="action-button"
        onActionButtonClick={onActionButtonClick}
      />
    );

    const closeButton = screen.queryByRole('button', {
      name: 'action-button',
      hidden: true,
    });
    await userEvent.click(closeButton);
    expect(onActionButtonClick).toHaveBeenCalledTimes(1);
  });

  it('interpolates matching className based on kind prop', () => {
    const { rerender, container } = render(
      <Callout title="Notification title" />
    );
    const kinds = ['info', 'warning'];
    kinds.forEach((kind) => {
      rerender(<Callout title="Notification title" kind={kind} />);
      expect(container.firstChild).toHaveClass(
        `${prefix}--actionable-notification--${kind}`
      );
    });
  });

  it('places low contrast class based on lowContrast prop', () => {
    const { container } = render(
      <Callout title="Notification title" lowContrast />
    );
    expect(container.firstChild).toHaveClass(
      `${prefix}--actionable-notification--low-contrast`
    );
  });

  it('allows non-interactive elements as children', () => {
    render(
      <Callout title="Notification title">
        <p>Sample text</p>
      </Callout>
    );
    // eslint-disable-next-line testing-library/prefer-presence-queries
    expect(screen.queryByText(/Sample text/i)).toBeInTheDocument();
  });

  it('allows interactive elements as children', () => {
    render(
      <Callout title="Notification title" titleId="interactive-test">
        <button aria-describedby="interactive-test">Sample text</button>
      </Callout>
    );
    // eslint-disable-next-line testing-library/prefer-presence-queries
    expect(screen.queryByText(/Sample text/i)).toBeInTheDocument();
  });

  it('supports `title`, `subtitle`, `titleId` props', () => {
    const { container } = render(
      <Callout title="A title" subtitle="A subtitle" titleId="aTitleID">
        <p>Sample text</p>
      </Callout>
    );
    // eslint-disable-next-line testing-library/prefer-presence-queries
    expect(screen.queryByText(/A title/i)).toBeInTheDocument();
    expect(screen.queryByText(/A subtitle/i)).toBeInTheDocument();
    expect(screen.queryByText(/A title/i)).toHaveAttribute('id', 'aTitleID');
  });

  it('enforces aria-describedby on interactive children elements', () => {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {});

    expect(() => {
      render(
        <Callout title="Notification title" titleId="titleId">
          <button type="button" aria-describedby="titleId">
            Sample button text
          </button>
        </Callout>
      );
    }).not.toThrow();

    expect(spy).not.toHaveBeenCalled();
    spy.mockRestore();
  });
});
