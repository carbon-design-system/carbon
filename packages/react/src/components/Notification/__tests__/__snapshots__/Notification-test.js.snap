// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionableNotification should render 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="actionable-notification-subtitle-:r1:"
    class="cds--actionable-notification cds--actionable-notification--toast cds--actionable-notification--error"
    role="alertdialog"
  >
    <span
      class="cds--visually-hidden"
      role="link"
      tabindex="0"
    >
      Focus sentinel
    </span>
    <div
      class="cds--actionable-notification__details"
    >
      <svg
        aria-hidden="true"
        class="cds--toast-notification__icon"
        fill="currentColor"
        focusable="false"
        height="20"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10,1c-5,0-9,4-9,9s4,9,9,9s9-4,9-9S15,1,10,1z M13.5,14.5l-8-8l1-1l8,8L13.5,14.5z"
        />
        <path
          d="M13.5,14.5l-8-8l1-1l8,8L13.5,14.5z"
          data-icon-path="inner-path"
          opacity="0"
        />
        <title>
          error icon
        </title>
      </svg>
      <div
        class="cds--actionable-notification__text-wrapper"
      >
        <div
          class="cds--actionable-notification__content"
        />
      </div>
    </div>
    <div
      class="cds--actionable-notification__button-wrapper"
    >
      <button
        class="cds--actionable-notification__action-button cds--btn cds--btn--sm cds--layout--size-sm cds--btn--tertiary"
        type="button"
      >
        My custom action
      </button>
      <button
        aria-label="close notification"
        class="cds--actionable-notification__close-button"
        title="close notification"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="cds--actionable-notification__close-icon"
          fill="currentColor"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          viewBox="0 0 32 32"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17.4141 16L24 9.4141 22.5859 8 16 14.5859 9.4143 8 8 9.4141 14.5859 16 8 22.5859 9.4143 24 16 17.4141 22.5859 24 24 22.5859 17.4141 16z"
          />
        </svg>
      </button>
    </div>
    <span
      class="cds--visually-hidden"
      role="link"
      tabindex="0"
    >
      Focus sentinel
    </span>
  </div>
</DocumentFragment>
`;

exports[`InlineNotification should render 1`] = `
<DocumentFragment>
  <div
    class="cds--inline-notification cds--inline-notification--error"
    role="status"
  >
    <div
      class="cds--inline-notification__details"
    >
      <svg
        aria-hidden="true"
        class="cds--inline-notification__icon"
        fill="currentColor"
        focusable="false"
        height="20"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10,1c-5,0-9,4-9,9s4,9,9,9s9-4,9-9S15,1,10,1z M13.5,14.5l-8-8l1-1l8,8L13.5,14.5z"
        />
        <path
          d="M13.5,14.5l-8-8l1-1l8,8L13.5,14.5z"
          data-icon-path="inner-path"
          opacity="0"
        />
        <title>
          error icon
        </title>
      </svg>
      <div
        class="cds--inline-notification__text-wrapper"
      >
        <div
          class="cds--inline-notification__title"
          dir="auto"
        >
          Notification title
        </div>
      </div>
    </div>
    <button
      aria-hidden="true"
      aria-label="close notification"
      class="cds--inline-notification__close-button"
      tabindex="-1"
      title="close notification"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="cds--inline-notification__close-icon"
        fill="currentColor"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.4141 16L24 9.4141 22.5859 8 16 14.5859 9.4143 8 8 9.4141 14.5859 16 8 22.5859 9.4143 24 16 17.4141 22.5859 24 24 22.5859 17.4141 16z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`ToastNotification should render 1`] = `
<DocumentFragment>
  <div
    class="cds--toast-notification cds--toast-notification--error"
    role="status"
  >
    <svg
      aria-hidden="true"
      class="cds--toast-notification__icon"
      fill="currentColor"
      focusable="false"
      height="20"
      preserveAspectRatio="xMidYMid meet"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10,1c-5,0-9,4-9,9s4,9,9,9s9-4,9-9S15,1,10,1z M13.5,14.5l-8-8l1-1l8,8L13.5,14.5z"
      />
      <path
        d="M13.5,14.5l-8-8l1-1l8,8L13.5,14.5z"
        data-icon-path="inner-path"
        opacity="0"
      />
      <title>
        error icon
      </title>
    </svg>
    <div
      class="cds--toast-notification__details"
    >
      <div
        class="cds--toast-notification__title"
        dir="auto"
      >
        Notification title
      </div>
    </div>
    <button
      aria-hidden="true"
      aria-label="close notification"
      class="cds--toast-notification__close-button"
      tabindex="-1"
      title="close notification"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="cds--toast-notification__close-icon"
        fill="currentColor"
        focusable="false"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.4141 16L24 9.4141 22.5859 8 16 14.5859 9.4143 8 8 9.4141 14.5859 16 8 22.5859 9.4143 24 16 17.4141 22.5859 24 24 22.5859 17.4141 16z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
